### PR TITLE
[WIP] Implement stained glass blocks and panes

### PR DESCRIPTION
**Pull request by lhochbaum**
This PR implements stained glass blocks and their respective panes.

### CODE REVIEW

File: dragonfly/block/glass_pane.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+// AABB adjusts bounding box of the glass pane.
+func (p GlassPane) AABB(pos world.BlockPos, w *world.World) []physics.AABB {
+	return calculateThinBounds(pos, w)
+}
+
+// ConnectsWith ...
+func (p GlassPane) ConnectsWith(other world.Block) bool {
+	switch other.(type) {
+	case GlassPane, StainedGlassPane: // TODO(lhochbaum): Add iron bars as soon as they're implemented.
```
Comment (Sandertv) [inline dragonfly/block/glass_pane.go:43]: Note that glass panes also connect with most other blocks. We probably need some interface to determine if a block has a specific solid face, where not implementing it by default means it is solid on all sides.

File: dragonfly/block/grass.go [hunk 2]
```diff
@@ -81,3 +81,8 @@ func (g Grass) EncodeBlock() (name string, properties map[string]interface{}) {
 	}
 	return "minecraft:grass", nil
 }
+
+// FaceSolidTo ...
+func (Grass) FaceSolidTo(_ world.BlockPos, _ world.Face, _ world.Block) bool {
```
Comment (Sandertv) [inline dragonfly/block/grass.go:86]: Note that grass is a grass block, not tall grass.

File: dragonfly/block/wood_slab.go [hunk 3]
```diff
@@ -163,6 +163,11 @@ func (s WoodSlab) SideClosed(pos, side world.BlockPos, w *world.World) bool {
 	return !s.Top && side[1] == pos[1]-1
 }
 
+// FaceSolidTo ...
+func (s WoodSlab) FaceSolidTo(_ world.BlockPos, _ world.Face, _ world.Block) bool {
```
Comment (Sandertv) [inline dragonfly/block/wood_slab.go:167]: This can be removed with the introduction of the block models.

File: dragonfly/block/wood_stairs.go [hunk 4]
```diff
@@ -113,6 +113,12 @@ func (s WoodStairs) SideClosed(pos, side world.BlockPos, w *world.World) bool {
 	return s.Model().FaceSolid(pos, pos.Face(side), w)
 }
 
+// FaceSolidTo ...
+func (s WoodStairs) FaceSolidTo(_ world.BlockPos, face world.Face, _ world.Block) bool {
```
Comment (Sandertv) [inline dragonfly/block/wood_stairs.go:117]: This can be removed with the introduction of the block models.

File: dragonfly/block/model/thin.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+		block := w.Block(pos)
+		if block.Model().FaceSolid(pos, i.Opposite(), w) {
+			boxes = append(boxes, mainBox.ExtendTowards(int(i), offset))
+		}
+	}
+	return append(boxes, mainBox)
+}
+
+// FaceSolid ...
+func (t Thin) FaceSolid(_ world.BlockPos, _ world.Face, _ *world.World) bool {
```
Comment (Sandertv) [inline dragonfly/block/model/thin.go:31]: You can remove the `_` completely if you use none of the parameters.

File: dragonfly/block/model/thin.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+	return false
```
Comment (Sandertv) [inline dragonfly/block/model/thin.go:32]: Thin blocks have a solid side at the bottom: You can hang lanterns from glass panes and iron bars, and farmland/grass paths turn into dirt when these are placed on top of them.

File: dragonfly/block/model/thin.go [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+func (t Thin) AABB(pos world.BlockPos, w *world.World) []physics.AABB {
+	const offset = 0.4375
+
+	boxes := make([]physics.AABB, 0, 5)
+	mainBox := physics.NewAABB(mgl64.Vec3{offset, 0, offset}, mgl64.Vec3{1 - offset, 1, 1 - offset})
+
+	for i := world.Face(2); i < 6; i++ {
+		pos := pos.Side(i)
+		block := w.Block(pos)
+		if block.Model().FaceSolid(pos, i.Opposite(), w) {
```
Comment (Sandertv) [inline dragonfly/block/model/thin.go:23]: This should also check if it's another `Thin` model and do the following code if so. Maybe also add a TODO to do the same for a future Wall model. (Glass panes and the likes also connect to, say, cobblestone walls)

File: dragonfly/block/block.go [hunk 8]
```diff
@@ -141,3 +141,10 @@ type transparent struct{}
 func (transparent) LightDiffusionLevel() uint8 {
 	return 0
 }
+
+// PartiallySolid represents a block that is not fully solid on every side or to every kind of block.
+type PartiallySolid interface {
```
Comment (Sandertv) [inline dragonfly/block/block.go:146]: This can be removed with the introduction of the block models.

File: config.toml [hunk 9]
```diff
@@ -14,8 +14,6 @@
   # The message shown to players when the server is shutting down. The message may be left empty to direct
   # players to the server list directly.
   ShutdownMessage = "Server closed."
-  # AuthEnabled controls whether or not players must be connected to Xbox Live in order to join the server.
-  AuthEnabled = true
```
Comment (Sandertv) [inline config.toml:18]: I think you committed this on accident.

File: dragonfly/block/stained_glass.go [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+	transparent
+	thin
+
+	// Colour specifies the colour of the block.
+	Colour colour.Colour
+}
+
+// BreakInfo ...
+func (g StainedGlass) BreakInfo() BreakInfo {
+	return BreakInfo{
```
Comment (T14Raptor) [inline dragonfly/block/stained_glass.go:21]: Incorrect formatting here.

Comment (Sandertv) [reply dragonfly/block/stained_glass.go:21]: I think this is fine actually.

Comment (T14Raptor) [reply dragonfly/block/stained_glass.go:21]: Oh alright.

File: dragonfly/block/model/thin.go [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+		if _, isThin := block.Model().(Thin); isThin || block.Model().FaceSolid(pos, i.Opposite(), w) {
+			boxes = append(boxes, mainBox.ExtendTowards(int(i), offset))
+		}
+	}
+	return append(boxes, mainBox)
+}
+
+// FaceSolid ...
+func (t Thin) FaceSolid(_ world.BlockPos, face world.Face, _ *world.World) bool {
+	return face < 2
```
Comment (Sandertv) [inline dragonfly/block/model/thin.go:34]: I don't think the top side is solid, you can't put torches on it. Just check if `face == world.FaceDown`

File: dragonfly/block/block.go (hunk 8 above)
Comment (Sandertv) [reply dragonfly/block/block.go:146]: Don't forget this one.

**Pull request merged**